### PR TITLE
FEATURE: TMA-1597 Logging lcm execution result

### PR DESCRIPTION
**Pull request by hung-nguyen-hoang**

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/132/ : SUCCESS in 8s

File: bin/run_brick.rb [hunk 1]
```diff
@@ -14,6 +14,27 @@ def get_brick_params(prefix)
   ENV.select { |k,| k.to_s.match(/^#{prefix}.*/) }.map { |k, v| [k.slice(prefix.length..-1), v] }.to_h
 end
 
+def log_execution_result(params, log, brick_type, error)
```
Comment (danh-ung) [inline bin/run_brick.rb]: log_execution_result --> log_execution_summary_error

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/134/ : SUCCESS in 7s

File: bin/run_brick.rb (hunk 1 above)
Comment (hung-nguyen-hoang) [reply bin/run_brick.rb]: Done fix it.

File: lib/gooddata/core/logging.rb [hunk 2]
```diff
@@ -23,7 +23,7 @@ module GoodData
   DEFAULT_SPLUNKLOG_OUTPUT = STDERR
 
   class << self
-    attr_accessor :logger, :rest_logger, :gd_logger
+    attr_accessor :logger, :rest_logger, :gd_logger, :lcm_logger
```
Comment (phong-nguyen-duy) [inline lib/gooddata/core/logging.rb]: lcm_logger is not a logger. It is just a error message. We should not keep it in this global class.

File: bin/run_brick.rb [hunk 3]
```diff
@@ -31,11 +50,14 @@ def get_brick_params(prefix)
...
+  params['GDC_EXECUTION_LOG_PATH'] = ENV['GDC_EXECUTION_LOG_PATH']
+  params['GDC_EXECUTION_RESULT_LOG_PATH'] = ENV['GDC_EXECUTION_RESULT_LOG_PATH']
   log.info "action=#{brick_type}_execution status=start commit_hash=#{commit_hash} execution_id=#{execution_id}"
   brick_pipeline.call(params)
 rescue Exception => e # rubocop:disable RescueException
-	execution_log = GoodData.logger
-	execution_log.error "Execution failed. Error: #{e}" unless execution_log.nil?
+  execution_log = GoodData.logger
+  execution_log.error "Execution failed. Error: #{e}" unless execution_log.nil?
+  log_execution_summary_error(params, log, brick_type, e)
```
Comment (phong-nguyen-duy) [inline bin/run_brick.rb]: should not reuse the 'params' because it was modified when execute pipeline.

File: lib/gooddata/core/logging.rb (hunk 2 above)
Comment (hung-nguyen-hoang) [reply lib/gooddata/core/logging.rb]: I cannot find any where good enough to store the error message. So could you please suggest me where we can store it ?
Thank you.

Comment (phong-nguyen-duy) [reply lib/gooddata/core/logging.rb]: I think we can define a new custom error (ex: LcmError) that contains all errors. 
In lcm.rb we create this error and throws it to outside. 
In run_brick.rb, we catch the LcmError and handle the error messages

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/136/ : SUCCESS in 8s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/137/ : SUCCESS in 6s

File: lib/gooddata/core/logging.rb (hunk 2 above)
Comment (hung-nguyen-hoang) [reply lib/gooddata/core/logging.rb]: Done fix it

File: bin/run_brick.rb (hunk 3 above)
Comment (hung-nguyen-hoang) [reply bin/run_brick.rb]: Won't fix. We only get parameters that don't modified during execute pipeline.

File: lib/gooddata/bricks/middleware/logger_middleware.rb [hunk 4]
```diff
@@ -36,7 +36,7 @@ def call(params)
           log_directory = params['GDC_LOG_DIRECTORY']
           execution_id = params['GDC_EXECUTION_ID']
           FileUtils.mkpath log_directory
-          logger = Logger.new("#{log_directory}/#{execution_id}.log")
+          logger = params['GDC_EXECUTION_LOG_PATH'].nil? ? Logger.new("#{log_directory}/#{execution_id}.log") : params['GDC_EXECUTION_LOG_PATH']
```
Comment (phong-nguyen-duy) [inline lib/gooddata/bricks/middleware/logger_middleware.rb]: why assign directly params['GDC_EXECUTION_LOG_PATH'] to a logger? Is is a String or a logger?

File: bin/run_brick.rb (hunk 3 above)
Comment (phong-nguyen-duy) [reply bin/run_brick.rb]: If you look into the method convert_params in this code base https://github.com/gooddata/gooddata-ruby/blob/master/lib/gooddata/lcm/lcm2.rb#L164, you will see that the params has been modified by the symbolize_keys! method

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/142/ : SUCCESS in 7s

File: bin/run_brick.rb [hunk 5]
```diff
@@ -14,28 +14,65 @@ def get_brick_params(prefix)
   ENV.select { |k,| k.to_s.match(/^#{prefix}.*/) }.map { |k, v| [k.slice(prefix.length..-1), v] }.to_h
 end
 
+def log_execution_summary_error(params, log, brick_type, error_message)
+  result = {
+    errorCode: 'lcm.error',
+    message: error_message,
+    parameters: []
+  }
+
```
Comment (danh-ung) [inline bin/run_brick.rb]: 1. JSON should have root element executionResult, status=OK or ERROR. In the future we can extend it for WARN status.
{
  "executionResult": {
    "status": "OK", // ERROR
    "message": "Blabla",
  }
}

2. log_execution_summary_error --> log_execution_result(status, message)

File: bin/run_brick.rb [hunk 6]
```diff
@@ -14,28 +14,65 @@ def get_brick_params(prefix)
...
+  begin
+    log_directory = params['GDC_LOG_DIRECTORY']
+    execution_id = params['GDC_EXECUTION_ID']
+    execution_result_logger_file = params['GDC_EXECUTION_RESULT_LOG_PATH'].nil? ? "#{log_directory}/#{execution_id}_result.json" : params['GDC_EXECUTION_RESULT_LOG_PATH']
+    File.open(execution_result_logger_file, 'w') { |file| file.write(JSON.pretty_generate(result)) }
+    log.info "action=log_#{brick_type}_result execution_id=#{execution_id} status=success error_message=#{error_message}"
```
Comment (danh-ung) [inline bin/run_brick.rb]: Remove this log, it's not necessary

File: bin/run_brick.rb [hunk 7]
```diff
@@ -14,28 +14,65 @@ def get_brick_params(prefix)
...
+  rescue Exception => e # rubocop:disable RescueException
+    log.info "action=log_#{brick_type}_result execution_id=#{execution_id} status=failed exception=#{e}"
```
Comment (danh-ung) [inline bin/run_brick.rb]: exception should be log as warn at least, log.warn ...

File: bin/run_brick.rb [hunk 8]
```diff
@@ -14,28 +14,65 @@ def get_brick_params(prefix)
...
+  end
+end
+
+def handle_error(params, log, brick_type, error, error_message)
+  execution_log = GoodData.logger
+  execution_log.error "Execution failed. Error: #{error}" unless execution_log.nil?
+  log_execution_summary_error(params, log, brick_type, error_message)
+  log.info "action=#{brick_type}_execution status=failed commit_hash=#{params['GOODDATA_RUBY_COMMIT']} execution_id=#{params['GDC_EXECUTION_ID']} exception=#{error}"
```
Comment (danh-ung) [inline bin/run_brick.rb]: log.error

File: bin/run_brick.rb [hunk 9]
```diff
@@ -14,28 +14,65 @@ def get_brick_params(prefix)
...
 begin
+  commit_hash = ENV['GOODDATA_RUBY_COMMIT'] || ''
+  execution_id = ENV['GDC_EXECUTION_ID']
+  log_directory = ENV['GDC_LOG_DIRECTORY'] || '/tmp/'
+  error_log_params = {
+    GOODDATA_RUBY_COMMIT: commit_hash,
+    GDC_EXECUTION_ID: execution_id,
+    GDC_LOG_DIRECTORY: log_directory,
+    GDC_EXECUTION_LOG_PATH: params['GDC_EXECUTION_LOG_PATH'],
+    GDC_EXECUTION_RESULT_LOG_PATH: params['GDC_EXECUTION_RESULT_LOG_PATH']
```
Comment (phong-nguyen-duy) [inline bin/run_brick.rb]: the params is not initialized

File: bin/run_brick.rb (hunk 5 above)
Comment (danh-ung) [reply bin/run_brick.rb]: 3. Define enum for status OK, ERROR

File: bin/run_brick.rb [hunk 10]
```diff
@@ -14,28 +14,65 @@ def get_brick_params(prefix)
...
-  execution_id = ENV['GDC_EXECUTION_ID']
   params['gooddata_ruby_commit'] = commit_hash
-  params['GDC_LOG_DIRECTORY'] = ENV['GDC_LOG_DIRECTORY'] || '/tmp/'
+  params['GDC_LOG_DIRECTORY'] = log_directory
   params['GDC_EXECUTION_ID'] = execution_id
+  params['GDC_EXECUTION_LOG_PATH'] = ENV['GDC_EXECUTION_LOG_PATH']
+  params['GDC_EXECUTION_RESULT_LOG_PATH'] = ENV['GDC_EXECUTION_RESULT_LOG_PATH']
+
   log.info "action=#{brick_type}_execution status=start commit_hash=#{commit_hash} execution_id=#{execution_id}"
   brick_pipeline.call(params)
```
Comment (danh-ung) [inline bin/run_brick.rb]: We should update execution result for OK case too, update_execution_result(OK, "")

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/143/ : SUCCESS in 7s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/146/ : SUCCESS in 36s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/147/ : SUCCESS in 7s

**Comment by phong-nguyen-duy** [comment]
LGTM

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-merge-compliance-check/401/ : SUCCESS in 25s

**Pull request merged**